File: AZ-204_Develop_Message_Based_Solutions/code/interact.py
from azure.servicebus import ServiceBusClient, ServiceBusMessage
from azure.eventhub import EventHubProducerClient, EventData
from azure.eventgrid import EventGridPublisherClient, EventGridEvent

# Interacting with Service Bus
servicebus_conn_str = "<your_servicebus_connection_string>"
servicebus_queue_name = "<your_servicebus_queue_name>"

servicebus_client = ServiceBusClient.from_connection_string(servicebus_conn_str)
with servicebus_client:
    servicebus_sender = servicebus_client.get_queue_sender(queue_name=servicebus_queue_name)
    message = ServiceBusMessage("Hello, Service Bus!")
    servicebus_sender.send_messages(message)

# Interacting with Event Hub
eventhub_conn_str = "<your_eventhub_connection_string>"
eventhub_name = "<your_eventhub_name>"

eventhub_client = EventHubProducerClient.from_connection_string(eventhub_conn_str, eventhub_name=eventhub_name)
with eventhub_client:
    event_data = EventData("Hello, Event Hub!")
    eventhub_client.send(event_data)

# Interacting with Event Grid
eventgrid_endpoint = "<your_eventgrid_endpoint>"
eventgrid_key = "<your_eventgrid_key>"
eventgrid_topic = "<your_eventgrid_topic>"

eventgrid_client = EventGridPublisherClient(eventgrid_endpoint, credential=eventgrid_key)
event = EventGridEvent(
    data={
        "message": "Hello, Event Grid!"
    },
    subject="MySubject",
    event_type="MyEventType",
    data_version="1.0"
)
eventgrid_client.send_events(event)

File: AZ-204_Develop_Message_Based_Solutions/code/interact.cs
using Azure.Messaging.ServiceBus;
using Azure.Messaging.EventHubs;
using Azure.Messaging.EventGrid;

// Interacting with Service Bus
string connectionString = "<your-service-bus-connection-string>";
string queueName = "<your-queue-name>";

await using (ServiceBusClient client = new ServiceBusClient(connectionString))
{
    ServiceBusSender sender = client.CreateSender(queueName);
    ServiceBusReceiver receiver = client.CreateReceiver(queueName);

    // Send a message
    ServiceBusMessage message = new ServiceBusMessage("Hello, Service Bus!");
    await sender.SendMessageAsync(message);

    // Receive a message
    ServiceBusReceivedMessage receivedMessage = await receiver.ReceiveMessageAsync();
    string messageBody = receivedMessage.Body.ToString();
    Console.WriteLine($"Received message: {messageBody}");
}

// Interacting with Event Hub
string eventHubConnectionString = "<your-event-hub-connection-string>";
string eventHubName = "<your-event-hub-name>";

await using (EventHubProducerClient producer = new EventHubProducerClient(eventHubConnectionString, eventHubName))
{
    // Send an event
    EventData eventData = new EventData(Encoding.UTF8.GetBytes("Hello, Event Hub!"));
    await producer.SendAsync(eventData);
}

await using (EventHubConsumerClient consumer = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, eventHubConnectionString, eventHubName))
{
    // Receive events
    await foreach (PartitionEvent partitionEvent in consumer.ReadEventsAsync())
    {
        string eventData = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
        Console.WriteLine($"Received event: {eventData}");
    }
}

// Interacting with Event Grid
string topicEndpoint = "<your-event-grid-topic-endpoint>";
string topicKey = "<your-event-grid-topic-key>";

EventGridPublisherClient client = new EventGridPublisherClient(new Uri(topicEndpoint), new AzureKeyCredential(topicKey));

// Publish an event
EventGridEvent eventGridEvent = new EventGridEvent
{
    Id = Guid.NewGuid().ToString(),
    EventType = "MyApp.Events.CustomEvent",
    Subject = "MyApp/Orders",
    DataVersion = "1.0",
    Data = new { OrderId = 123, Product = "ABC" }
};

await client.SendEventAsync(eventGridEvent);